### ISSUE TITLE
will you provide the training script?

### ISSUE THREAD

**Issue by gaopeng-eugene** [hszhao/PSPNet#12]

**Comment by dongzhuoyao** [comment]
same

**Comment by mjohn123** [comment]
I am also regrading to the question. Thanks

**Comment by RickyWong33** [comment]
same

**Comment by huaxinxiao** [comment]
Has anyone re-trained successfully?

**Comment by rener1199** [comment]
same

**Comment by justinbuzzni** [comment]
same

**Comment by hszhao** [comment]
Hi, for the training, the issues are mainly related to bn layer:
1. Whether to update the parameters (mean,variance,slope,bias) of 'bn'?
    -If you are working on the same (or similar) dataset as the released model, you can just fix the 'bn' layer for fine tuning; if not, you may need to update the parameters.
2. What may need more attention when updating parameters?
    -The batch size when doing batch normalization is important and it's better to set it above 16 in a calculation step for we need to keep the current mean and variance approximate to the global statistics that will be used in the testing step. While semantic segmentation is memory consuming and to maintain a larger crop size (related to certain dataset) may cause small batch size on each GPU card. So during our training step, we use MPI to gather data from different GPU cards and then do the bn operation. While it seems that current official Caffe doesn't support such communication. We are trying to make our training code compatible with BVLC and you can have a glance at Caffe vision of [yjxiongIf](https://github.com/yjxiong/caffe/tree/mem) which is a OpenMPI-based Multi-GPU version. If you are working on other datasets, maybe other platform can support such bn communication. Sorry for the inconvenient of the current released evaluation code.
    Thanks.

**Comment by mjohn123** [comment]
@hszhao : Thanks for your information. I am working in same cityscapes dataset. I am using 1 GPU TitanX Pascal. Is it possible to run your training model in my computer? If not, could you reduce the Resnet depth layer such as 54? I also a beginner of caffer, so I do not know how can I make the training model from your first point

**Comment by Fromandto** [comment]
@hszhao I am training a 713 resolution pspnet on 2 x 12gb titan x with batch size 1, and it seems almost all memories are used. 

So I guess training with batchsize 16 would require about 32 titan x cards (12gb memory) ?

I cannot find details about how many gpus are used in the paper, so I want to confirm that how many gpus are required to train with batchsize 16 according to your experience ?

I really wonder what is the quantitative performance improvement between batchsize 16 and  batchsize 1, because in the paper and this thread you emphasize that batchsize matters yet in deeplab-v2 (and according to my own experience) training with batchsize 1 also works (to some extent). Do I really need to use batchsize 16 (and potentially 32 cards ?) to achieve ideal performance ? ...

**Comment by huaxinxiao** [comment]
@Fromandto  If your batchsize is 1, the batch normalization layer may be not work. However, the bn layer seems important to the performance of PSPNet.

**Comment by Fromandto** [comment]
@huaxinxiao  yes, this is exactly what i am concerned ... but I just don't have 32 gpus (or is there anything wrong with my setting so that 4 gpus are enough to train 16 batch ?)

**Comment by huaxinxiao** [comment]
@Fromandto  Smaller crop size (<321) will work in 4 gpus. Besides, you should use [OpenMPI-based Multi-GPU caffe](https://github.com/yjxiong/caffe/tree/mem) to gather  the bn parameters.

**Comment by suhyung** [comment]
@Fromandto Could you share your training script?

**Comment by Fromandto** [comment]
@suhyung I am using the training script of deeplab-v2. it is compatible.

**Comment by SoonminHwang** [comment]
@Fromandto @hszhao Could you tell me some details for training? I'm using deeplab-v2 caffe, and I'm ready to train a model with my own python script. But I don't have any proper initial weights for pspnet101-VOC2012.prototxt. I tried to use the initial parameters from [deeplab-v2](http://liangchiehchen.com/projects/DeepLabv2_resnet.html), the layer names are different. Should I train a network for pre-trained model on ImageNet by myself?

**Comment by qizhuli** [comment]
@SoonminHwang you can either transfer the weights over by matcaffe/pycaffe, or you can replace the resnet part of the pspnet prototxt with the deeplab version.

**Comment by ThienAnh** [comment]
@SoonminHwang  Did you have init weights file?

**Comment by tkasarla** [comment]
What init weights should we use for training the cityscapes model for pspnet?

**Comment by melody-rain** [comment]
@huaxinxiao can you train this pspnet with SyncBN?

**Comment by holyseven** [comment]
I've implemented sync batch normalization in pure tensorflow, which makes possible to train and reproduce the performance of PSPNet: [batch norm across GPUs](https://github.com/holyseven/PSPNet-TF-Reproduce/blob/master/model/utils_mg.py#L177).
